### ISSUE TITLE
Allow users to store data on the Zcash blockchain

### ISSUE THREAD

**Issue by jackgavigan** [zcash/zcash#2573]
People already use OP_RETURN as a means to store data on the Zcash blockchain (e.g. t1UU7zbHH7PhFg2nu3AbxmmSgEAZstfmVaR), and data can be attached to shielded transactions using the encrypted memo field.

Why don't we allow people to store (a limited amount) of data on the Zcash blockchain in a way that allows the data to be updated (like a variable), and allows the data to be easily referenced?

**Comment by bitcartel** [comment]
Sounds like data streams:
https://www.multichain.com/developers/data-streams/

**Comment by jackgavigan** [comment]
@bitcartel: Yes, that's pretty much exactly what I was thinking about (although I think it would be better to store large datasets off-chain, and just store a URI and a hash on-chain), and I think the moniker is apt.

**Comment by nathan-at-least** [comment]
Let's refine the requirements by answering these questions:

- who can publish data?
- who can read published data?

For example, for the case of "anyone who knows a z-addr can publish data to that z-addr and anyone who knows the z-addrs viewkey can read that data", then we already have that with stock encrypted memo fields.

I'm wondering what the most flexible definitions of publishers/receivers are. I think it might be something like this:

- there are an arbitrary number of "channels" and anyone may create a channel to produce a "publication capability" and a "subscription capability", each of which has a standard encoding into an opaque byte sequence.
- anyone with access to a channel's publication capability can append arbitrary data to the channel
- anyone with access to that same channel's subscription capability can see all data appended
- neither capability implies the other (so if you *only* have a channel's publish capability, you cannot subscribe, and vice versa)

It seems with those primitives, you can construct arbitrary sets of publishers/subscribers for a variety of applications. Some useful cases:

a. broadcast a channel's publication capability to the world, but keep the subscription capability secret. This behaves just like a published z-addr (pub cap) and incoming-only viewkey (sub cap).
b. broadcast a channel's subscription capability to the world, but keep the publication capability secret. This is a 1 to many broadcast channel.

**Comment by nathan-at-least** [comment]
Let's refine the requirements by answering these questions:

- who can publish data?
- who can read published data?

For example, for the case of "anyone who knows a z-addr can publish data to that z-addr and anyone who knows the z-addrs viewkey can read that data", then we already have that with stock encrypted memo fields.

I'm wondering what the most flexible definitions of publishers/receivers are. I think it might be something like this:

- there are an arbitrary number of "channels" and anyone may create a channel to produce a "publication capability" and a "subscription capability", each of which has a standard encoding into an opaque byte sequence.
- anyone with access to a channel's publication capability can append arbitrary data to the channel
- anyone with access to that same channel's subscription capability can see all data appended
- neither capability implies the other (so if you *only* have a channel's publish capability, you cannot subscribe, and vice versa)

It seems with those primitives, you can construct arbitrary sets of publishers/subscribers for a variety of applications. Some useful cases:

a. broadcast a channel's publication capability to the world, but keep the subscription capability secret. This behaves just like a published z-addr (pub cap) and incoming-only viewkey (sub cap).
b. broadcast a channel's subscription capability to the world, but keep the publication capability secret. This is a 1 to many broadcast channel.
